File: src/net/minecraft/client/render/block/entity/SignBlockEntityRenderer.cpp
#include "net/minecraft/client/render/block/entity/SignBlockEntityRenderer.hpp"
#include "net/minecraft/block/Block.hpp"
#include "net/minecraft/block/entity/SignBlockEntity.hpp"
#include "net/minecraft/client/font/TextRenderer.hpp"
#include "net/minecraft/client/render/RenderCore.hpp"
#include "net/minecraft/client/render/RenderType.hpp"
namespace net::minecraft::client::render::block::entity {
void SignBlockEntityRenderer::render(
    const net::minecraft::block::entity::BlockEntity& blockEntity, double x, double y, double z, float tickDelta) {
 (void)tickDelta;
 const auto* sign = dynamic_cast<const net::minecraft::block::entity::SignBlockEntity*>(&blockEntity);
 if(sign == nullptr) {
  return;
 }
 net::minecraft::block::Block* block = sign->getBlock();
 render::core::modelViewStack().push();
 const float scale = 0.6666667f;
 if(block == net::minecraft::block::Block::SIGN) {
  render::core::modelViewStack().translate(
      static_cast<float>(x) + 0.5f, static_cast<float>(y) + 0.75f * scale, static_cast<float>(z) + 0.5f);
  const float yaw = static_cast<float>(sign->getPushedBlockData() * 360) / 16.0f;
  render::core::modelViewStack().rotate(-yaw, 0.0f, 1.0f, 0.0f);
  model.stick.visible = true;
 } else {
  const int meta = sign->getPushedBlockData();
  float yaw = 0.0f;
  if(meta == 2) {
   yaw = 180.0f;
  } else if(meta == 4) {
   yaw = 90.0f;
  } else if(meta == 5) {
   yaw = -90.0f;
  }
  render::core::modelViewStack().translate(
      static_cast<float>(x) + 0.5f, static_cast<float>(y) + 0.75f * scale, static_cast<float>(z) + 0.5f);
  render::core::modelViewStack().rotate(-yaw, 0.0f, 1.0f, 0.0f);
  render::core::modelViewStack().translate(0.0f, -0.3125f, -0.4375f);
  model.stick.visible = false;
 }
 bindTexture("/item/sign.png");
 render::core::modelViewStack().push();
 render::core::modelViewStack().scale(scale, -scale, -scale);
 model.render();
 render::core::modelViewStack().pop();
 font::TextRenderer* textRenderer = getTextRenderer();
 if(textRenderer == nullptr) {
  render::core::modelViewStack().pop();
  return;
 }
 const float textScale = 0.016666668f * scale;
 render::core::modelViewStack().translate(0.0f, 0.5f * scale, 0.07f * scale);
 render::core::modelViewStack().scale(textScale, -textScale, textScale);
 render::RenderPassScope passScope(render::RenderType::text());
 render::core::enableDepthTest();
 render::core::depthMask(false);
 const int color = 0;
 for(int i = 0; i < 4; ++i) {
  std::string line = sign->texts[static_cast<std::size_t>(i)];
  if(i == sign->currentRow) {
   line = "> " + line + " <";
  }
  textRenderer->draw(
      line, -textRenderer->getWidth(line) / 2, i * 10 - static_cast<int>(sign->texts.size()) * 5, color);
 }
 render::core::modelViewStack().pop();
}
} // namespace net::minecraft::client::render::block::entity

File: mcp/src/net/minecraft/client/render/block/entity/SignBlockEntityRenderer.java
/*
 * Decompiled with CFR 0.1.1 (FabricMC 57d88659).
 * 
 * Could not load the following classes:
 *  net.fabricmc.api.EnvType
 *  net.fabricmc.api.Environment
 */
package net.minecraft.client.render.block.entity;

import net.fabricmc.api.EnvType;
import net.fabricmc.api.Environment;
import net.minecraft.block.Block;
import net.minecraft.block.entity.SignBlockEntity;
import net.minecraft.client.font.TextRenderer;
import net.minecraft.client.render.block.entity.BlockEntityRenderer;
import net.minecraft.client.render.block.entity.SignModel;
import org.lwjgl.opengl.GL11;

@Environment(value=EnvType.CLIENT)
public class SignBlockEntityRenderer
extends BlockEntityRenderer {
    private SignModel model = new SignModel();

    public void render(SignBlockEntity arg, double d, double e, double f, float g) {
        float f2;
        Block block = arg.getBlock();
        GL11.glPushMatrix();
        float f3 = 0.6666667f;
        if (block == Block.SIGN) {
            GL11.glTranslatef((float)d + 0.5f, (float)e + 0.75f * f3, (float)f + 0.5f);
            float f4 = (float)(arg.getPushedBlockData() * 360) / 16.0f;
            GL11.glRotatef(-f4, 0.0f, 1.0f, 0.0f);
            this.model.stick.visible = true;
        } else {
            int n = arg.getPushedBlockData();
            f2 = 0.0f;
            if (n == 2) {
                f2 = 180.0f;
            }
            if (n == 4) {
                f2 = 90.0f;
            }
            if (n == 5) {
                f2 = -90.0f;
            }
            GL11.glTranslatef((float)d + 0.5f, (float)e + 0.75f * f3, (float)f + 0.5f);
            GL11.glRotatef(-f2, 0.0f, 1.0f, 0.0f);
            GL11.glTranslatef(0.0f, -0.3125f, -0.4375f);
            this.model.stick.visible = false;
        }
        this.bindTexture("/item/sign.png");
        GL11.glPushMatrix();
        GL11.glScalef(f3, -f3, -f3);
        this.model.render();
        GL11.glPopMatrix();
        TextRenderer textRenderer = this.getTextRenderer();
        f2 = 0.016666668f * f3;
        GL11.glTranslatef(0.0f, 0.5f * f3, 0.07f * f3);
        GL11.glScalef(f2, -f2, f2);
        GL11.glNormal3f(0.0f, 0.0f, -1.0f * f2);
        GL11.glDepthMask(false);
        int n = 0;
        for (int i = 0; i < arg.texts.length; ++i) {
            String string = arg.texts[i];
            if (i == arg.currentRow) {
                string = "> " + string + " <";
                textRenderer.draw(string, -textRenderer.getWidth(string) / 2, i * 10 - arg.texts.length * 5, n);
                continue;
            }
            textRenderer.draw(string, -textRenderer.getWidth(string) / 2, i * 10 - arg.texts.length * 5, n);
        }
        GL11.glDepthMask(true);
        GL11.glColor4f(1.0f, 1.0f, 1.0f, 1.0f);
        GL11.glPopMatrix();
    }
}

